### PR TITLE
[mdex-integration][feat] : liquidate strategy

### PR DESCRIPTION
**Pull request by abc-alpaca**

### CODE REVIEW

File: contracts/6/protocol/strategies/mdex-restricted/MdexRestrictedStrategyLiquidate.sol [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+contract MdexRestrictedStrategyLiquidate is OwnableUpgradeSafe, ReentrancyGuardUpgradeSafe, IStrategy {
+  using SafeToken for address;
+  using SafeMath for uint256;
+
+  IMdexFactory public factory;
+  IMdexRouter public router;
+
+  mapping(address => bool) public okWorkers;
+
+  address public mdx;
```
Comment (jr-alpaca) [inline contracts/6/protocol/strategies/mdex-restricted/MdexRestrictedStrategyLiquidate.sol:38]: declare mdx after router in line 35

File: contracts/6/protocol/strategies/mdex-restricted/MdexRestrictedStrategyLiquidate.sol [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+  /// @notice require that only allowed workers are able to do the rest of the method call
+  modifier onlyWhitelistedWorkers() {
+    require(okWorkers[msg.sender], "MdexRestrictedStrategyLiquidate::onlyWhitelistedWorkers:: bad worker");
+    _;
+  }
+
+  /// @dev Create a new liquidate strategy instance.
+  /// @param _router The IMdexRouter smart contract.
+  function initialize(IMdexRouter _router, address _mdx) external initializer {
```
Comment (jr-alpaca) [inline contracts/6/protocol/strategies/mdex-restricted/MdexRestrictedStrategyLiquidate.sol:48]: please add comment for _mdx initialize agrument

File: test/Mdex_Restricted_StrategyLiquidate.test.ts [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+import { deploy } from "@openzeppelin/hardhat-upgrades/dist/utils";
+chai.use(solidity);
+const { expect } = chai;
+
+const FOREVER = "2000000000";
+const MDX_PER_BLOCK ="51600000000000000000"
+
+describe("MdexRestricted_StrategyLiquidate",()=>{
+
+// FIXME do declare variable
```
Comment (jr-alpaca) [inline test/Mdex_Restricted_StrategyLiquidate.test.ts:17]: remove this comment

File: contracts/6/protocol/strategies/mdex-restricted/MdexRestrictedStrategyLiquidate.sol [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  function setWorkersOk(address[] calldata workers, bool isOk) external onlyOwner {
+    for (uint256 idx = 0; idx < workers.length; idx++) {
+      okWorkers[workers[idx]] = isOk;
+    }
+  }
+
+  function withdrawTradingRewards(address to) external onlyOwner {
+    SwapMining(router.swapMining()).takerWithdraw();
+    SafeToken.safeTransfer(mdx, to, SafeToken.myBalance(mdx));
+  }
```
Comment (jr-alpaca) [inline contracts/6/protocol/strategies/mdex-restricted/MdexRestrictedStrategyLiquidate.sol:101]: add getMiningRewards()

File: test/Mdex_Restricted_StrategyLiquidate.test.ts [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+      expect(
+        bobBTokenAfter.sub(bobBTokenBefore),
+        "Bob's balance should increase by 1.499374217772215269 BTOKEN"
+      ).to.be.bignumber.eq(ethers.utils.parseEther("1").add(ethers.utils.parseEther("0.499374217772215269")));
+      expect(await lp.balanceOf(strat.address)).to.be.bignumber.eq(ethers.utils.parseEther("0"));
+      expect(await lp.balanceOf(await bob.getAddress())).to.be.bignumber.eq(ethers.utils.parseEther("0"));
+      expect(await baseToken.balanceOf(lp.address)).to.be.bignumber.eq(ethers.utils.parseEther("0.500625782227784731"));
+      expect(await farmingToken.balanceOf(lp.address)).to.be.bignumber.eq(ethers.utils.parseEther("0.2"));
+      // formula is blockReward*poolAlloc/totalAlloc = 825600000000000000000 *100/100
+      expect(mdexTokenAfter.sub(mdexTokenBefore)).to.be.bignumber.eq("877200000000000000000");
```
Comment (jr-alpaca) [inline test/Mdex_Restricted_StrategyLiquidate.test.ts:360]: assert mdx amount change with getMiningRewards()

File: contracts/6/protocol/strategies/mdex-restricted/MdexRestrictedStrategyLiquidate.sol [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+  /// @notice get Mining reward
+  /// @param pids pool IDs of all to be get rewards,
```
Comment (jr-alpaca) [inline contracts/6/protocol/strategies/mdex-restricted/MdexRestrictedStrategyLiquidate.sol:103]: /// @param pids pool IDs of all to be get rewards.
small typo

File: contracts/6/protocol/strategies/mdex-restricted/MdexRestrictedStrategyLiquidate.sol [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+pragma solidity 0.6.6;
+
+import "@openzeppelin/contracts-ethereum-package/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts-ethereum-package/contracts/utils/ReentrancyGuard.sol";
+import "@openzeppelin/contracts-ethereum-package/contracts/math/Math.sol";
+
+import "../../apis/mdex/IMdexRouter.sol";
+import "../../apis/mdex/IMdexFactory.sol";
+import "../../apis/mdex/SwapMining.sol";
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/mdex-restricted/MdexRestrictedStrategyLiquidate.sol:22]: Create `ISwapMining.sol` and import it instead

File: contracts/6/protocol/strategies/mdex-restricted/MdexRestrictedStrategyLiquidate.sol [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+      okWorkers[workers[idx]] = isOk;
+    }
+  }
+
+  function withdrawTradingRewards(address to) external onlyOwner {
+    SwapMining(router.swapMining()).takerWithdraw();
+    SafeToken.safeTransfer(mdx, to, SafeToken.myBalance(mdx));
+  }
+
+  /// @notice get Mining reward
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/mdex-restricted/MdexRestrictedStrategyLiquidate.sol:102]: `/// @dev Get mining rewards by pIds`

File: contracts/6/protocol/strategies/mdex-restricted/MdexRestrictedStrategyLiquidate.sol [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+  /// @param pids pool IDs of all to be get rewards.
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/mdex-restricted/MdexRestrictedStrategyLiquidate.sol:103]: `/// @param pool ids to retrieve reward amount.`

**Review by spicysquid168 — CHANGES_REQUESTED** [review]

**Review by spicysquid168 — APPROVED** [review]

**Pull request merged**